### PR TITLE
Improve error message when export encounters non-local input

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106403

Previously, you would get an error like

```
Dynamo input and output is a strict subset of traced input/output
```

now you get

```
Cannot export model which references tensors that are neither
buffers/parameters/constants nor are direct inputs.  For each tensor, if you'd
like this tensor to be an explicit input, add it as a dummy argument
to the top-level model definition you are exporting; if you would
like its value to be embedded as an exported constant, wrap its access
in a function marked with @assume_constant_result.

G['bulbous_bouffant'], accessed at:
  File "test_export.py", line N, in f
    return bulbous_bouffant + y
```

This doesn't handle outputs, I'm going to hit that next.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -1066,9 +1066,50 @@ def result_capturing_wrapper(*graph_inputs):
...
+
+        if input_errors:
+            raise UserError(
+                UserErrorType.INVALID_INPUT,
+                "Cannot export model which references tensors that are neither "
+                "buffers/parameters/constants nor are direct inputs.  For each tensor, if you'd "
+                "like this tensor to be an explicit input, add it as a dummy argument "
+                "to the top-level model definition you are exporting; if you would "
+                "like its value to be embedded as an exported constant, wrap its access "
+                "in a function marked with @assume_constant_result.\n\n"
```
Comment (tugsbayasgalan) [inline torch/_dynamo/eval_frame.py:1109]: Are we ok with advertizing @assume_constant_result?

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -1266,7 +1289,26 @@ def remove_node(self, node):
...
+        # In eager, we are generally OK with adding graph inputs whenever we
+        # want, because we take care of writing the bytecode that knows how
+        # to source all the inputs.
+        #
+        # In export, this is bad, because you want a self-contained export
+        # object which only depends on the inputs you explicitly passed to it.
+        # So we are a bit more strict about what sources can become inputs
+        # in export
+        if self.export_root:
+            if not is_from_local_source(source, allow_cell_or_freevar=False):
```
Comment (tugsbayasgalan) [inline torch/_dynamo/output_graph.py:1307]: nit: `is_from_local_source(source, allow_cell_or_freevar=(not self.export_root))`

**Review by tugsbayasgalan — APPROVED** [review]

File: torch/_dynamo/eval_frame.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/eval_frame.py:1109]: I am happy with removing it! But it seemed like a useful tool for certain situations?

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/output_graph.py:1307]: no, I don't want to do this test at all when not self.export_root

File: torch/_dynamo/variables/builder.py [hunk 3]
```diff
@@ -894,25 +894,30 @@ def wrap_tensor(self, value: torch.Tensor):
             ) or is_traceable_wrapper_subclass(value), type(value)
             ignore_subclass = False
 
+        # NB: this just says we accessed a tensor from the same source again
+        # (e.g., a tensor lives in a global foo, and we LOAD_GLOBAL it twice).
+        # This is distinct from two distinct sources mapping to the same
+        # Tensor (per id())!  No guard is necessary here.  See below for the
+        # other case.
         is_duplicate_tensor = source in self.tx.output.input_source_to_var
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:902]: This is an unrelated code comment that I added for clarity about this flow haha

**Pull request closed without merging**